README
   segatex-ng daemon name segatexd - console program for SELinux 

   In 2006, segatex was born to help SELinux system administration.  
   It is written with C++ and Qt3 originally.  Although segatex is maintained untill 2017, 
   it became harder to catch up the new Qt version.  
   In 2017, the author, Shintaro Fujiwara, decided to write console program of segatex, 
   and started search for good base program.  "segatex-ng" means "segatex next generation".  
   He found "nss-pam-ldapd" written by Arthur de Jong of West Consuling is a excellent program, 
   so decided to borrow codes from it.  
   Although segatex-ng is not concerned about LDAP, its code style is the ones of the original program.  

   segatex-ng is a program for sysadmins using SELinux in everyday work.  
   segatexd is segatex-ng's daemon name.

   Copyright (C) 2017 Shintaro Fujiwara 

   This software is free software; you can redistribute it and/or
   modify it under the terms of the GNU Lesser General Public
   License as published by the Free Software Foundation; either
   version 2.1 of the License, or (at your option) any later version.

   This library is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   Lesser General Public License for more details.

   You should have received a copy of the GNU Lesser General Public
   License along with this library; if not, write to the Free Software
   Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
   02110-1301 USA

HOW TO BUILD
==================
segatexd uses autoconf and automake for building. Installing
segatexd should be as simple as:

  $ ./configure [options]
  $ make
  # make install

It is a good idea to first go through the options of configure by running:

  $ ./configure --help

INSTALLATION
============

  # make install

UINSTALLATION
============

  # make uninstall

HOW TO USE 
==================

  # segatexd -d

  Only root can use this program.

CONFIGURATION
=============

  /etc/segatexd.conf

  File permission is set 600.

SENDING SIGNALS 
==================
Executing the next command will reload the configuration file.

# kill -s HUP `pidof segatexd`

Executing the next command will stop the program.

# kill -s INT `pidof segatexd`

REPORTING BUGS
==============

  Email to <shintaro.fujiwara@gmai.com>


NEWS
changes from 0.1.4 to 0.1.5
---------------------------
* Added logging functionality 
* Code cleaning 
* Fixed properly handle SIGINT  
* Added avoid_oom_killer function 

changes from 0.1.3 to 0.1.4
---------------------------
* Fixed getopt functionality in main.c 
* Changed src directory to server directory 
* Added syslog messages 
* Added to link with libaudit 
* Fixed config file read functionality 

changes from 0.1.2 to 0.1.3
---------------------------
* SIGHUP functionality was added 

changes from 0.1.1 to 0.1.2
---------------------------
* daemonized process will be checked by strings

changes from 0.1.0 to 0.1.1
---------------------------
* Now segatexd is daemonized 
* configuration file name was changed to segatexd.conf from segatex-ng.conf 



server/main.c
/*
   main.c - main file for segatex-ng.
   This file contains the contents of segatex-ng.

   Copyright (C) 2017 Shintaro Fujiwara

   This library is free software; you can redistribute it and/or
   modify it under the terms of the GNU Lesser General Public
   License as published by the Free Software Foundation; either
   version 2.1 of the License, or (at your option) any later version.

   This library is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   Lesser General Public License for more details.

   You should have received a copy of the GNU Lesser General Public
   License along with this library; if not, write to the Free Software
   Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
   02110-1301 USA
*/

#include <stdio.h>
#include <errno.h>
#include <signal.h>
#include <stdlib.h>
#include <unistd.h>
#include <syslog.h>
#include <getopt.h>
#include <string.h>
#include <fcntl.h>
#include <sys/stat.h>
#include <sys/types.h>
#include "main.h"
#include "../segatexd.h"
#include "message.c"
#include "daemonize.c"

/* brief Print help for this application */

void print_help(void)
{
    printf("\n Usage: %s [OPTIONS]\n\n", app_name);
    printf("  Options:\n");
    //printf("   -p --pid_file  filename   PID file used by daemonized app\n");
    printf("   -t --test_conf filename   Test configuration file\n");
    printf("   -r --reload               Reload the configuration file\n");
    printf("   -s --stop                 Stop segatexd daemon program\n");
    printf("   -d --daemon               Daemonize this application\n");
    printf("   -h --help                 Print this help\n");
    //printf("   -c --conf_file filename   Read configuration from the file\n");
    //printf("   -l --log_file  filename   Write logs to the file\n");
    printf("\n");
}

/*avoid oom killer */

static void avoid_oom_killer(void)
{
    int oomfd, len, rc;
    char *score = NULL;

    /* New kernels use different technique */	
    if ((oomfd = open("/proc/self/oom_score_adj",
                O_NOFOLLOW | O_WRONLY)) >= 0) {
        score = "-1000";
    } else if ((oomfd = open("/proc/self/oom_adj",
                O_NOFOLLOW | O_WRONLY)) >= 0) {
        score = "-17";
    } else {
        segatex_msg(LOG_NOTICE, "Cannot open out of memory adjuster");
        return;
    }

    len = strlen(score);
    rc = write(oomfd, score, len);
    //segatex_msg(LOG_INFO,"len:%d",len);
    //segatex_msg(LOG_INFO,"rc:%d",rc);
    if (rc != len)
        segatex_msg(LOG_NOTICE, "Unable to adjust out of memory score");

    close(oomfd);
}

/* Main function */

int main(int argc, char *argv[])
{
    //app_name = argv[0];
    app_name = "segatexd";
    //setting message_mode
    set_sgmessage_mode(MSG_SYSLOG, DBG_NO); 

    static struct option long_options[] = {
        //{"pid_file", required_argument, 0, 'p'},
        {"test_conf", required_argument, 0, 't'},
        {"reload", no_argument, 0, 'r'},
        {"stop", no_argument, 0, 's'},
        {"daemon", no_argument, 0, 'd'},
        {"help", no_argument, 0, 'h'},
        {NULL, 0, 0, 0}
    };

    int value, option_index = 0, ret;
    char *log_file_name = NULL;
    int start_daemonized = 0;
    int duplicate_value = 0;
    int LEN = 4096;
    char line[LEN];

    /* Try to process all command line arguments */
    while ((value = getopt_long(argc, argv, "trsdh::", long_options, &option_index)) != -1) {
        switch (value) {
        //   case 'p':
        //        pid_file_name = strdup(optarg);
        //        break;
            case 't':
                cfg_init(fname);
                break;
            case 'r':
                //printf("duplicate_value:%d\n",duplicate_value);
                if ( duplicate_value == 1 ){
                    printf("-r, -s and -d could not be selected at the same time\n");
                    return EXIT_FAILURE; 
                }
                duplicate_value = 1;
                cfg_init(fname);
                handle_signal(SIGHUP);
                break;
            case 's':
                if ( duplicate_value == 1 ){
                    printf("-r, -s and -d could not be selected at the same time\n");
                    return EXIT_FAILURE; 
                }
                duplicate_value = 1;
                running =3;
                log_stream = stdout;
                kill(read_pid(),SIGINT);
                break;
            case 'd':
                //printf("duplicate_value:%d\n",duplicate_value);
                if ( duplicate_value == 1 ){
                    printf("-r, -s and -d could not be selected at the same time\n");
                    return EXIT_FAILURE; 
                }
                duplicate_value = 1;
                start_daemonized = 1;
                break;
            case 'h':
                print_help();
                return EXIT_SUCCESS;
                case '?':
                print_help();
                return EXIT_FAILURE;
            default:
                break;
        }
    }

    /* Tell kernel not to kill us */
    avoid_oom_killer();

    int SIG_VALUE;
    printf("SIG_VALUE is %d\n",SIG_VALUE);

    /* Daemon will handle two signals */
    signal(SIGINT, handle_signal);
    signal(SIGHUP, handle_signal);

    int i;
    int i2;
    int fd2;

    i = is_selinux_enabled();
    i2 = audit_is_enabled(1);
    if (i = 1)
        printf("SELinux is enabled.\n");
    else
        printf("SELinux is not enabled.\n");
    if (i2 = 1)
        printf("Audit is enabled.\n");
    else if (i2 = 0)
        printf("Audit is not enabled.\n");
    else if (i2 = -1)
        printf("Audit got error.\n");

    /* SIGINT should have running as 3*/
    if (running != 3)
    {
        /* Initialize the conf file reading procedure. */
        cfg_init(fname);

        /* When daemonizing is requested at command line. */
        if (start_daemonized == 1) {
            /* It is also possible to use glibc function deamon()
             * at this point, but it is useful to customize your daemon. */
            daemonize();
        }

        /* Open system log and write message to it */
        openlog(argv[0], LOG_PID|LOG_CONS, LOG_DAEMON);
        segatex_msg(LOG_INFO, "Started %s", app_name);

        /* Try to open log file to this daemon */
        if (log_file_name != NULL) {
            log_stream = fopen(log_file_name, "a+");
            if (log_stream == NULL) {
                segatex_msg(LOG_ERR, "Can not open log file: %s, error: %s",
                    log_file_name, strerror(errno));
                log_stream = stdout;
            }
        } else {
            log_stream = stdout;
        }

        /* This global variable can be changed in function handling signal */
        running = 1;
    }

    /* Never ending loop of server */
    while (running == 1) {
        /* Debug print */
        ret = fprintf(log_stream, "Debug: %d\n", counter++);
        if (ret < 0) {
            segatex_msg(LOG_ERR, "Can not write to log stream: %s, error: %s",
                (log_stream == stdout) ? "stdout" : log_file_name, strerror(errno));
            break;
        }
        ret = fflush(log_stream);
        if (ret != 0) {
            segatex_msg(LOG_ERR, "Can not fflush() log stream: %s, error: %s",
                (log_stream == stdout) ? "stdout" : log_file_name, strerror(errno));
            break;
        }
        /* TODO: dome something useful here */

        /* Real server should use select() or poll() for waiting at
         * asynchronous event. Note: sleep() is interrupted, when
         * signal is received. */
        segatex_msg(LOG_INFO, "this is segatexd-server");
        //for debug
        //segatex_msg(LOG_INFO, "PID of segatexd is %d",read_pid());
        //for debug
        sleep(delay);
    }

    /* Close log file, when it is used. */
    if (log_stream != stdout) {
        fclose(log_stream);
    }

    /* Write system log and close it. */
    segatex_msg(LOG_INFO, "Stopped %s", app_name);
    closelog();

    /* unlink pid file */
    if (pid_file_name != NULL)
        unlink(pid_file_name);

    return EXIT_SUCCESS;
}
